# DigiCow Farmer Training Adoption Challenge

### Loading data and Simple EDA

In [2]:
# import libraries 

import pandas as pd 
import numpy as np 
import ast
import re

In [3]:
# load datasets

train = pd.read_csv("D:/digicow-farmer-training-adoption-challenge20260210-32396-19ubatl (1)/Train.csv")
test = pd.read_csv("D:/digicow-farmer-training-adoption-challenge20260210-32396-19ubatl (1)/Test.csv")
prior = pd.read_csv("D:/digicow-farmer-training-adoption-challenge20260210-32396-19ubatl (1)/Prior.csv")

print("Train shape:", train.shape)
print("Test shape:", test.shape)
print("Prior shape:", prior.shape)


Train shape: (13536, 17)
Test shape: (5621, 14)
Prior shape: (44882, 17)


In [4]:
# view columns and first rows

print("Train columns:\n", train.columns.tolist())
print("\nTrain head:")
display(train.head())

print("Prior columns:\n", prior.columns.tolist())
print("\nPrior head:\n")
display(prior.head())

Train columns:
 ['ID', 'farmer_name', 'training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'adopted_within_07_days', 'adopted_within_90_days', 'adopted_within_120_days', 'has_topic_trained_on', 'trainer', 'topics_list']

Train head:


ID  farmer_name training_day  gender registration       age  \
0  ID_CENCC8  FAR_eqbhscj   2024-01-03  Female       Manual  Above 35   
1  ID_YTO0FF  FAR_qlwtyik   2024-01-03  Female       Manual  Above 35   
2  ID_1476PE  FAR_somfzxp   2024-01-03  Female       Manual  Above 35   
3  ID_MLKLIR  FAR_ongcqyd   2024-01-03  Female       Manual  Above 35   
4  ID_V5ZVTA  FAR_ztsbhhm   2024-01-03  Female         Ussd  Below 35   

    group_name  belong_to_cooperative      county    subcounty          ward  \
0  GRP_yvpakgc                      0  CNT_lpotuu  SUB_lpotuuf  WRD_lpotuufh   
1  GRP_zemrbsy                      1  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   
2  GRP_zmblxsw                      0  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   
3  GRP_psdrfni                      0  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   
4  GRP_yvpakgc                      0  CNT_lpotuu  SUB_lpotuuf  WRD_lpotuufh   

   adopted_within_07_days  adopted_within_90_days  adopted_within_120_days  \
0                       0                       0                        0   
1                       0                       0                        0   
2                       0                       0                        0   
3                       0                       0                        0   
4                       0                       0                        0   

   has_topic_trained_on           trainer  \
0                     0  ['TRA_szrwyfzz']   
1                     1  ['TRA_rkvyofbh']   
2                     1  ['TRA_rkvyofbh']   
3                     1  ['TRA_rkvyofbh']   
4                     0  ['TRA_szrwyfzz']   

                                         topics_list  
0                 [['Ndume App', 'Poultry Feeding']]  
1         [['Poultry Housing'], ['Poultry Housing']]  
2  [['Asili Fertilizer (Organic)', 'Biosecurity I...  
3  [['Poultry Products'], ['Record Keeping In Dai...  
4                 [['Ndume App', 'Poultry Feeding']]

Prior columns:
 ['ID', 'farmer_name', 'training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'adopted_within_07_days', 'adopted_within_90_days', 'adopted_within_120_days', 'has_topic_trained_on', 'trainer', 'topics_list']

Prior head:



ID  farmer_name training_day  gender registration       age  \
0  ID_70GP6F  FAR_leopgvh   2024-01-03  Female       Manual  Above 35   
1  ID_IWQOWJ  FAR_vdcjfxm   2024-01-03  Female         Ussd  Above 35   
2  ID_Z3ES85  FAR_hfkybdg   2024-01-03  Female       Manual  Above 35   
3  ID_JNZM6R  FAR_hfkybdg   2024-01-03  Female       Manual  Above 35   
4  ID_BNJ1GU  FAR_hfkybdg   2024-01-03  Female       Manual  Above 35   

    group_name  belong_to_cooperative      county    subcounty          ward  \
0  GRP_yvpakgc                      0  CNT_lpotuu  SUB_lpotuuf  WRD_lpotuufh   
1  GRP_yvpakgc                      0  CNT_lpotuu  SUB_lpotuuf  WRD_lpotuufh   
2  GRP_zmblxsw                      0  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   
3  GRP_psdrfni                      0  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   
4  GRP_psdrfni                      1  CNT_fhdsoy  SUB_mdyljqn  WRD_atkhhvon   

   adopted_within_07_days  adopted_within_90_days  adopted_within_120_days  \
0                       0                       0                        0   
1                       0                       0                        0   
2                       0                       0                        0   
3                       0                       0                        0   
4                       0                       0                        0   

   has_topic_trained_on       trainer  \
0                     0  TRA_szrwyfzz   
1                     0  TRA_szrwyfzz   
2                     1  TRA_rkvyofbh   
3                     1  TRA_rkvyofbh   
4                     1  TRA_rkvyofbh   

                                         topics_list  
0                   ['Ndume App', 'Poultry Feeding']  
1                   ['Ndume App', 'Poultry Feeding']  
2  ['Asili Fertilizer (Organic)', 'Biosecurity In...  
3                               ['Poultry Products']  
4                        ['Record Keeping In Dairy']

In [5]:
# Clean column names (strip spaces)

def clean_columns(df):
    df.columns = df.columns.str.strip()
    return df

train = clean_columns(train)
test = clean_columns(test)
prior = clean_columns(prior)

print("Cleaned train columns:\n", train.columns.tolist())
print("Cleaned test columns:\n", test.columns.tolist())
print("Cleaned prior columns:\n", prior.columns.tolist())

Cleaned train columns:
 ['ID', 'farmer_name', 'training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'adopted_within_07_days', 'adopted_within_90_days', 'adopted_within_120_days', 'has_topic_trained_on', 'trainer', 'topics_list']
Cleaned test columns:
 ['ID', 'farmer_name', 'training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'has_topic_trained_on', 'trainer', 'topics_list']
Cleaned prior columns:
 ['ID', 'farmer_name', 'training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'adopted_within_07_days', 'adopted_within_90_days', 'adopted_within_120_days', 'has_topic_trained_on', 'trainer', 'topics_list']


In [6]:
# Parse training_day as a date

for df, name in [(train, "train"), (test, "test"), (prior, "prior")]:
    if "training_day" in df.columns:
        df["training_day"] = pd.to_datetime(df["training_day"], errors = "coerce")
        print(name, "training_day parsed. Missing dates: ", df["training_day"].isna().sum())
    else:
        print(name, "does not have training_day column")   

train training_day parsed. Missing dates:  0
test training_day parsed. Missing dates:  0
prior training_day parsed. Missing dates:  0


In [7]:
# Target monotonicity check (7 ⟶ 90 ⟶ 120)
# Logical rule: If adopted_within_07_days == 1, then adopted_within_90_days and adopted_within_120_days should also be 1.

t7 = "adopted_within_07_days"
t90 = "adopted_within_90_days"
t120 = "adopted_within_120_days"

if all(col in train.columns for col in [t7, t90, t120]):
    violations_7_90 = ((train[t7] == 1) & (train[t90] == 0)).sum()
    violations_7_120 = ((train[t7] == 1) & (train[t120] == 0)).sum()
    violations_90_120 = ((train[t90] == 1) & (train[t120] == 0)).sum()

    print("violations (7->90) = ", violations_7_90)
    print("violations (7->120) = ", violations_7_120)
    print("violations (90->120) = ", violations_90_120)

else:
    print("Target column names not found. Check train.columns")


violations (7->90) =  0
violations (7->120) =  0
violations (90->120) =  0


In [8]:
# Missing value report (overall + top missing columns)

missing = train.isna().mean().sort_values(ascending=False)*100
missing = missing[missing > 0]


print("Columns with missing values (%)")
display(missing.to_frame("Missing_%").head(20))

Columns with missing values (%)


Empty DataFrame
Columns: [Missing_%]
Index: []

In [9]:
# Cardinality check (how many unique values)
# High-cardinality categoricals can overfit. Check key columns.

cols_to_check = ['farmer_name', 'gender', 'registration', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'trainer', 'age']

for col in cols_to_check:
    if col in train.columns:
        print(f"{col:15s} unique values:", train[col].nunique(dropna=False))

farmer_name     unique values: 13536
gender          unique values: 2
registration    unique values: 2
group_name      unique values: 864
belong_to_cooperative unique values: 2
county          unique values: 8
subcounty       unique values: 26
ward            unique values: 66
trainer         unique values: 12
age             unique values: 2


1) What the cardinalities imply

**farmer_name: 13,536 (very high)**

-This is basically “almost one per row” (since Train is ~14k).

-Do NOT use farmer_name as a feature → the model will memorize individuals and overfit.

Do use it for:

-grouping in CV (GroupKFold)

-joining/aggregating Prior history

**group_name: 864 (high-ish)**

-This likely represents training cohort / group.

-It can be predictive (some groups might have better follow-up), but also can cause leakage if the same group repeats across train/test.

-With CatBoost it’s fine to include, but validate with a grouped CV.

**Geography is small (excellent!)**

-county: 8, subcounty: 26, ward: 66

-This is great news: geography is informative but not too high-cardinality.

-can safely include them as categoricals.

**Trainer is tiny: trainer: 12 (very strong signal)**

-With only 12 trainers, this feature will likely be one of the most predictive. Keep it.

-can even make trainer×topic interaction features later.

**Simple demographics**

-gender: 2, registration: 2, belong_to_cooperative: 2, age: 2

-These are easy binary/low-cardinality predictors.

-Key takeaway: data is not heavy-cardinality except farmer_name and group_name. That makes modeling easier and more stable.

2) What this suggests about Prior dataset usefulness

With only:

-12 trainers

-66 wards

-864 groups

…adoption may be driven heavily by:

-training quality differences (trainer)

-group follow-up/coordination (group_name)

-farmer’s prior engagement (Prior history)

So Prior will be a big lever

3) Suggested modeling decisions based on this output

A) Feature inclusion

**Include as features:**

-gender, registration, belong_to_cooperative, age

-county, subcounty, ward

-trainer

-group_name (yes, but watch leakage)

**Exclude as feature:**

-farmer_name

B) CV strategy (very important)

Because farmer_name is nearly unique per row, GroupKFold by farmer won’t change much.

But GroupKFold by group_name is very important, because:

-If the same group appears across folds, the model can partially “memorize” group behavior.

-If test set contains similar group patterns, it’s ok; if not, it causes CV/LB mismatch.

So:

-Start with StratifiedKFold (baseline)

-Then validate with GroupKFold on group_name

-If scores drop a lot in grouped CV, rely more on grouped CV for truth

In [10]:
# Class imbalance (positive rates for each target)

for col in [t7, t90, t120]:
    if col in train.columns:
        pos_rate = train[col].mean()*100
        pos_count = train[col].sum()
        print(f"{col}: {pos_rate:.2f}% positives ({int(pos_count)} positive cases)")

adopted_within_07_days: 1.13% positives (153 positive cases)
adopted_within_90_days: 1.58% positives (214 positive cases)
adopted_within_120_days: 2.23% positives (302 positive cases)


Targets are extremely imbalanced:

- 98–99% of farmers DO NOT adopt

- Only ~150–300 positive examples exist per target

In [11]:
# Check if Prior actually contains multiple events per farmer

if "farmer_name" in prior.columns:
    events_per_farmer = prior.groupby("farmer_name").size()
    print("Prior: farmers:", events_per_farmer.shape[0])
    print("Prior: average events/farmer:", events_per_farmer.mean())
    print("Prior: max events for one farmer:", events_per_farmer.max())
    display(events_per_farmer.describe())
else:
    print("Prior does not contain farmer name column.")

Prior: farmers: 6719
Prior: average events/farmer: 6.679863074862331
Prior: max events for one farmer: 96


count    6719.000000
mean        6.679863
std         8.140113
min         1.000000
25%         2.000000
50%         4.000000
75%         8.000000
max        96.000000
dtype: float64

### Parsing and cleaning tricky columns

In [12]:
# Parse and clean the trainer column

train["trainer"].head(10)

0    ['TRA_szrwyfzz']
1    ['TRA_rkvyofbh']
2    ['TRA_rkvyofbh']
3    ['TRA_rkvyofbh']
4    ['TRA_szrwyfzz']
5    ['TRA_rkvyofbh']
6    ['TRA_rkvyofbh']
7    ['TRA_szrwyfzz']
8    ['TRA_szrwyfzz']
9    ['TRA_szrwyfzz']
Name: trainer, dtype: str

In [13]:
def parse_trainer(x):

    """
    Converts trainer field to:
      - trainer_list: a list of trainer IDs
      - trainer_primary: first trainer or 'Unknown'
      - n_trainers: number of trainers
    Handles cases where x is:
      - already a string like 'TRA_xxx'
      - a string representation of a list like "['TRA_xxx']"
      - missing (NaN)
    """

    if pd.isna(x):
        return [], "Unknown", 0

    if isinstance(x, str):
        x = x.strip()

        # If list-like string
        if x.startswith("[") and x.endswith("]"):
            try:
                trainers = ast.literal_eval(x)
            except:
                trainers = []
        else:
            trainers = [x]
    else:
        trainers = [str(x)]

    trainers = [t.strip() for t in trainers if isinstance(t, str) and t.strip() != ""]

    trainer_primary = trainers[0] if len(trainers) > 0 else "Unknown"
    n_trainers = len(trainers)

    return trainers, trainer_primary, n_trainers


In [14]:
# Apply parser to Train/Test/Prior

def add_trainer_features(df):
    trainer_lists = []
    trainer_primary = []
    n_trainers = []

    for x in df["trainer"]:
        t_list, t_primary, n_t = parse_trainer(x)
        trainer_lists.append(t_list)
        trainer_primary.append(t_primary)
        n_trainers.append(n_t)

    df = df.copy()
    df["trainer_list"] = trainer_lists
    df["trainer_primary"] = trainer_primary
    df["n_trainers"] = n_trainers
    return df

train = add_trainer_features(train)
test = add_trainer_features(test)
prior = add_trainer_features(prior)

train[['trainer', 'trainer_primary', 'n_trainers']].head(10)
                            


trainer trainer_primary  n_trainers
0  ['TRA_szrwyfzz']    TRA_szrwyfzz           1
1  ['TRA_rkvyofbh']    TRA_rkvyofbh           1
2  ['TRA_rkvyofbh']    TRA_rkvyofbh           1
3  ['TRA_rkvyofbh']    TRA_rkvyofbh           1
4  ['TRA_szrwyfzz']    TRA_szrwyfzz           1
5  ['TRA_rkvyofbh']    TRA_rkvyofbh           1
6  ['TRA_rkvyofbh']    TRA_rkvyofbh           1
7  ['TRA_szrwyfzz']    TRA_szrwyfzz           1
8  ['TRA_szrwyfzz']    TRA_szrwyfzz           1
9  ['TRA_szrwyfzz']    TRA_szrwyfzz           1

In [15]:
prior[['trainer', 'trainer_primary', 'n_trainers']].head(10)

trainer trainer_primary  n_trainers
0  TRA_szrwyfzz    TRA_szrwyfzz           1
1  TRA_szrwyfzz    TRA_szrwyfzz           1
2  TRA_rkvyofbh    TRA_rkvyofbh           1
3  TRA_rkvyofbh    TRA_rkvyofbh           1
4  TRA_rkvyofbh    TRA_rkvyofbh           1
5  TRA_rkvyofbh    TRA_rkvyofbh           1
6  TRA_rkvyofbh    TRA_rkvyofbh           1
7  TRA_rkvyofbh    TRA_rkvyofbh           1
8  TRA_rkvyofbh    TRA_rkvyofbh           1
9  TRA_rkvyofbh    TRA_rkvyofbh           1

In [16]:
train["trainer_primary"].value_counts().head(20)

trainer_primary
TRA_gertumxc    3049
TRA_suiifsur    2190
TRA_hyodnntj    2168
TRA_szrwyfzz    2028
TRA_rkvyofbh    1877
TRA_ubcgvofe    1456
TRA_kkzpfdtu     734
TRA_twwcfcum      24
TRA_dttdgplk      10
Name: count, dtype: int64

In [17]:
prior["n_trainers"].value_counts()

n_trainers
1    44882
Name: count, dtype: int64

In [18]:
prior["trainer_primary"].value_counts().head(20)


trainer_primary
TRA_gertumxc    19656
TRA_suiifsur     8834
TRA_rkvyofbh     7537
TRA_kkzpfdtu     5284
TRA_ubcgvofe     2991
TRA_hyodnntj      366
TRA_szrwyfzz      207
TRA_dttdgplk        4
TRA_twwcfcum        3
Name: count, dtype: int64

In [19]:
# Parse topics_list

def parse_topics(x):
    """"
    converts topic_list field to a clean list of topics.
    Handles:
        -string representation of list
        -actual list
        -missing values
    """

    # Handle missing at the top level
    if pd.isna(x):
        return []

    # If it's already a list
    if isinstance(x, list):
        topics = x

    # If it's a string (most common case)
    elif isinstance(x, str):
        s = x.strip()
        if s.startswith("[") and s.endswith("]"):
            try:
                topics = ast.literal_eval(s)
            except Exception:
                topics = []
        else:
            topics = [s]

    else:
        topics = [str(x)]

    # IMPORTANT: Flatten nested lists
    flat_topics = []
    for item in topics:
        if isinstance(item, list):
            flat_topics.extend(item)
        else:
            flat_topics.append(item)

    # Clean each topic safely
    clean = []
    for t in flat_topics:
        if isinstance(t, str):
            t = t.strip()
            if t != "":
                clean.append(t)

    return clean



In [20]:
# Add topic list + basic count features

def add_topic_features(df):
    df = df.copy()
    df["topics_parsed"] = df["topics_list"].apply(parse_topics)

    df["n_topics"] = df["topics_parsed"].apply(len)
    df["n_unique_topics"] = df["topics_parsed"].apply(lambda x: len(set(x)))
    df["topic_diversity"] = df["n_unique_topics"] / df["n_topics"].replace(0, np.nan)
    df["topic_diversity"] = df["topic_diversity"].fillna(0.0)

    return df

train = add_topic_features(train)
test  = add_topic_features(test)
prior = add_topic_features(prior)

train[["topics_list", "topics_parsed", "n_topics", "n_unique_topics", "topic_diversity"]].head(20)


topics_list  \
0                  [['Ndume App', 'Poultry Feeding']]   
1          [['Poultry Housing'], ['Poultry Housing']]   
2   [['Asili Fertilizer (Organic)', 'Biosecurity I...   
3   [['Poultry Products'], ['Record Keeping In Dai...   
4                  [['Ndume App', 'Poultry Feeding']]   
5   [['Poultry Products'], ['Record Keeping In Dai...   
6          [['Poultry Housing'], ['Poultry Housing']]   
7                  [['Ndume App', 'Poultry Feeding']]   
8                  [['Ndume App', 'Poultry Feeding']]   
9                  [['Ndume App', 'Poultry Feeding']]   
10  [['Benfits Of Sistema Biogas', 'Biosecurity In...   
11  [['Poultry Housing', 'Poultry Products', 'Reco...   
12                 [['Ndume App', 'Poultry Feeding']]   
13  [['Poultry Housing', 'Poultry Products', 'Reco...   
14  [['Poultry Products'], ['Record Keeping In Dai...   
15                             [['Microp+ Planting']]   
16  [['Microp+ Planting'], ['How To Succeed In Bre...   
17  [['Microp+ Planting'], ['Yara Maziwa Pro'], ['...   
18  [['Biosecurity In Poultry Farming', 'Calf Feed...   
19  [['How To Succeed In Breeding', 'Microp+ Plant...   

                                        topics_parsed  n_topics  \
0                        [Ndume App, Poultry Feeding]         2   
1                  [Poultry Housing, Poultry Housing]         2   
2   [Asili Fertilizer (Organic), Biosecurity In Po...        29   
3   [Poultry Products, Record Keeping In Dairy, Po...         3   
4                        [Ndume App, Poultry Feeding]         2   
5   [Poultry Products, Record Keeping In Dairy, Po...         3   
6                  [Poultry Housing, Poultry Housing]         2   
7                        [Ndume App, Poultry Feeding]         2   
8                        [Ndume App, Poultry Feeding]         2   
9                        [Ndume App, Poultry Feeding]         2   
10  [Benfits Of Sistema Biogas, Biosecurity In Pou...        13   
11  [Poultry Housing, Poultry Products, Record Kee...         6   
12                       [Ndume App, Poultry Feeding]         2   
13  [Poultry Housing, Poultry Products, Record Kee...         6   
14  [Poultry Products, Record Keeping In Dairy, Po...         3   
15                                 [Microp+ Planting]         1   
16  [Microp+ Planting, How To Succeed In Breeding,...         4   
17  [Microp+ Planting, Yara Maziwa Pro, Poultry Ma...         3   
18  [Biosecurity In Poultry Farming, Calf Feeding,...        12   
19  [How To Succeed In Breeding, Microp+ Planting,...         4   

    n_unique_topics  topic_diversity  
0                 2         1.000000  
1                 1         0.500000  
2                29         1.000000  
3                 3         1.000000  
4                 2         1.000000  
5                 3         1.000000  
6                 1         0.500000  
7                 2         1.000000  
8                 2         1.000000  
9                 2         1.000000  
10               12         0.923077  
11                3         0.500000  
12                2         1.000000  
13                3         0.500000  
14                3         1.000000  
15                1         1.000000  
16                4         1.000000  
17                3         1.000000  
18               12         1.000000  
19                4         1.000000

In [21]:
# Create top-k multi hot topic features

from collections import Counter

def get_top_topics (train_df, prior_df, k = 50):
    all_topics = []
    all_topics += train_df["topics_parsed"].sum() #sums lists together
    all_topics += prior_df["topics_parsed"].sum()

    topic_counts = Counter(all_topics)
    top_topics = [t for t, c in topic_counts.most_common(k)]
    return top_topics, topic_counts

TOP_K = 50
top_topics, topic_counts = get_top_topics(train, prior, k = TOP_K)

print("Top 10 Topics:")
print(top_topics[:10])

Top 10 Topics:
['The Benefits Of Ndume App', 'Herd Health Management.', 'Poultry Management Practises.', 'Poultry Health Management.', 'Poultry Management Practices', 'Biosecurity In Poultry Farming', 'Feeding A Lactating Cow', 'Calf Feeding', 'How To Succeed In Breeding', 'Reasons Why Ai Fails And Solutions']


In [22]:
# Add multi-hot columns to Train/test/prior

def add_multihot_topics(df, top_topics):
    df = df.copy()
    topic_set_list = df["topics_parsed"].apply(set)
    
    for t in top_topics:
        df[f"topic__{t}"] = topic_set_list.apply(lambda s: 1 if t in s else 0)

    return df

train = add_multihot_topics(train, top_topics)
prior = add_multihot_topics(prior, top_topics)
test = add_multihot_topics(test, top_topics)

# quick check
topic_cols = [c for c in train.columns if c.startswith("topic__")]
print("Number of topic multi-hot columns:", len(topic_cols))
train[topic_cols].head(20)

Number of topic multi-hot columns: 50


topic__The Benefits Of Ndume App  topic__Herd Health Management.  \
0                                  0                               0   
1                                  0                               0   
2                                  0                               0   
3                                  0                               0   
4                                  0                               0   
5                                  0                               0   
6                                  0                               0   
7                                  0                               0   
8                                  0                               0   
9                                  0                               0   
10                                 0                               0   
11                                 0                               0   
12                                 0                               0   
13                                 0                               0   
14                                 0                               0   
15                                 0                               0   
16                                 0                               0   
17                                 0                               0   
18                                 0                               0   
19                                 0                               0   

    topic__Poultry Management Practises.  topic__Poultry Health Management.  \
0                                      0                                  0   
1                                      0                                  0   
2                                      0                                  0   
3                                      0                                  0   
4                                      0                                  0   
5                                      0                                  0   
6                                      0                                  0   
7                                      0                                  0   
8                                      0                                  0   
9                                      0                                  0   
10                                     0                                  0   
11                                     0                                  0   
12                                     0                                  0   
13                                     0                                  0   
14                                     0                                  0   
15                                     0                                  0   
16                                     0                                  0   
17                                     0                                  0   
18                                     0                                  0   
19                                     0                                  0   

    topic__Poultry Management Practices  \
0                                     0   
1                                     0   
2                                     0   
3                                     0   
4                                     0   
5                                     0   
6                                     0   
7                                     0   
8                                     0   
9                                     0   
10                                    0   
11                                    0   
12                                    0   
13                                    0   
14                                    0   
15                                    0   
16                                    1   
17                                    1   
18            

In [23]:
# Confirm required columns exist and are correct

required_train_cols = ["farmer_name", "training_day", "trainer_primary", "topics_parsed" ]
required_prior_cols = ["farmer_name", "training_day", "trainer_primary", "topics_parsed" ]
required_test_cols = ["farmer_name", "training_day", "trainer_primary", "topics_parsed" ]

print("Missing in train:", [c for c in required_train_cols if c not in train.columns])
print("Missing in test:", [c for c in required_test_cols if c not in test.columns])
print("Missing in prior:", [c for c in required_prior_cols if c not in prior.columns])

# check dtype of training_day. Should be datetime
print("train training_dtype:", train["training_day"].dtype)
print("prior training_dtype:", prior["training_day"].dtype)
print("test training_dtype:", test["training_day"].dtype)

Missing in train: []
Missing in test: []
Missing in prior: []
train training_dtype: datetime64[us]
prior training_dtype: datetime64[us]
test training_dtype: datetime64[us]


**The leakage rule (IMPORTANT)**

We must ensure:

When creating features for a row at date D, we only use prior rows with date < D. We will enforce this with merge_asof(..., allow_exact_matches=False) which prevents using same-day rows.

In [24]:
# Sort data (required for merge_asof)

train = train.sort_values(["farmer_name", "training_day"]).reset_index(drop = True)
test = test.sort_values(["farmer_name", "training_day"]).reset_index(drop = True)
prior = prior.sort_values(["farmer_name", "training_day"]).reset_index(drop = True)

In [25]:
# Build “cumulative history” features inside Prior

# compute farmer history over time inside prior, then shift by 1 so an event doesn’t “see itself”.

prior_feat = prior.copy()

# 1) count event per farmer over time (cumulative)
prior_feat["prior_n_events_cum"] = prior_feat.groupby("farmer_name").cumcount() + 1

# 2) first training day per farmer
prior_feat["prior_first_day"] = prior_feat.groupby("farmer_name")["training_day"].transform("min")

# 3) previous training day per farmer
prior_feat["prior_prev_day"] = prior_feat.groupby("farmer_name")["training_day"].shift(1)

# 4) days since first or last at each prior row (inside prior timeline)
prior_feat["prior_days_since_first"] = (prior_feat["training_day"] - prior_feat["prior_first_day"]).dt.days
prior_feat["prior_days_since_last"] = (prior_feat["training_day"] - prior_feat["prior_prev_day"]).dt.days

prior_feat[["farmer_name","training_day","prior_first_day","prior_prev_day","prior_days_since_first","prior_days_since_last"]].head(20)

farmer_name training_day prior_first_day prior_prev_day  \
0   FAR_aaaqpow   2024-07-10      2024-07-10            NaT   
1   FAR_aaaqpow   2024-07-10      2024-07-10     2024-07-10   
2   FAR_aaaqpow   2024-07-10      2024-07-10     2024-07-10   
3   FAR_aaaqpow   2024-07-10      2024-07-10     2024-07-10   
4   FAR_aafjcve   2024-05-09      2024-05-09            NaT   
5   FAR_aahvays   2025-02-05      2025-02-05            NaT   
6   FAR_aajkwcm   2024-02-04      2024-02-04            NaT   
7   FAR_aajkwcm   2024-02-07      2024-02-04     2024-02-04   
8   FAR_aajkwcm   2024-02-07      2024-02-04     2024-02-07   
9   FAR_aajkwcm   2024-02-07      2024-02-04     2024-02-07   
10  FAR_aajkwcm   2024-02-07      2024-02-04     2024-02-07   
11  FAR_aakwhuh   2025-04-02      2025-04-02            NaT   
12  FAR_aakwhuh   2025-04-02      2025-04-02     2025-04-02   
13  FAR_aakwhuh   2025-04-02      2025-04-02     2025-04-02   
14  FAR_aakwhuh   2025-08-07      2025-04-02     2025-04-02   
15  FAR_aakwhuh   2025-08-07      2025-04-02     2025-08-07   
16  FAR_aalsnpt   2024-06-08      2024-06-08            NaT   
17  FAR_aalsnpt   2024-06-08      2024-06-08     2024-06-08   
18  FAR_aalsnpt   2024-06-08      2024-06-08     2024-06-08   
19  FAR_aalsnpt   2024-07-05      2024-06-08     2024-06-08   

    prior_days_since_first  prior_days_since_last  
0                        0                    NaN  
1                        0                    0.0  
2                        0                    0.0  
3                        0                    0.0  
4                        0                    NaN  
5                        0                    NaN  
6                        0                    NaN  
7                        3                    3.0  
8                        3                    0.0  
9                        3                    0.0  
10                       3                    0.0  
11                       0                    NaN  
12                       0                    0.0  
13                       0                    0.0  
14                     127                  127.0  
15                     127                    0.0  
16                       0                    NaN  
17                       0                    0.0  
18                       0                    0.0  
19                      27                   27.0

In [26]:
# Shift cumulative features to make them safe. We want features at time D to reflect history before D, not including D.

prior_feat["prior_n_events_before"] = prior_feat.groupby("farmer_name")["prior_n_events_cum"].shift(1)

# For "days since first" and "days since last", we want the values as of the previous event,
# but we’ll compute final versions after merging into train/test.

In [27]:
# Check if columns exist in prior

adopt_cols = ["adopted_within_07_days", "adopted_within_90_days", "adopted_within_120_days"]
print("Adopt columns in prior:", [c for c in adopt_cols if c in prior_feat.columns])


Adopt columns in prior: ['adopted_within_07_days', 'adopted_within_90_days', 'adopted_within_120_days']


In [28]:
# build cumulative sums and rates

for col in adopt_cols:
    if col in prior_feat.columns:
        # cumulative sum of adoptions
        prior_feat[f"{col}_cum_sum"] = prior_feat.groupby("farmer_name")[col].cumsum()
        # shift so it's "before current event"
        prior_feat[f"{col}_cum_sum_before"] = prior_feat.groupby("farmer_name")[f"{col}_cum_sum"].shift(1)

# adoption rates before current event = cum_adoptions_before / n_events_before
for col in adopt_cols:
    if col in prior_feat.columns:
        prior_feat[f"prior_rate_{col.split('_within_')[1]}"] = (
            prior_feat[f"{col}_cum_sum_before"] / prior_feat["prior_n_events_before"]
        )


In [29]:
# Merge Prior history into Train/Test (leakage-safe). use merge_asof to attach each train row to the farmer’s latest prior row strictly before that date.

# Choose which prior features to carry forward

keep_cols = [
    "farmer_name", "training_day",
    "prior_n_events_before",
    "prior_first_day",
    "prior_prev_day"
]

# Include rates if they exist
rate_cols = [c for c in prior_feat.columns if c.startswith("prior_rate_")]
keep_cols += rate_cols

prior_for_merge = prior_feat[keep_cols].sort_values(["farmer_name", "training_day"])


In [30]:
# Ensure datetime
train["training_day"] = pd.to_datetime(train["training_day"])
prior_for_merge["training_day"] = pd.to_datetime(prior_for_merge["training_day"])

# Correct sorting order for asof
train_left = train.sort_values(["training_day", "farmer_name"]).reset_index(drop=True)
prior_right = prior_for_merge.sort_values(["training_day", "farmer_name"]).reset_index(drop=True)

# merge for train
train_merged = pd.merge_asof(
    train_left,
    prior_right,
    by="farmer_name",
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)



In [31]:
# merge for test

test_left = test.sort_values(["training_day", "farmer_name"]).reset_index(drop=True)
test_left["training_day"] = pd.to_datetime(test_left["training_day"]).reset_index(drop = True)

test_merged = pd.merge_asof(
    test_left,
    prior_right,
    by="farmer_name",
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)

In [32]:
# Create final history features after merge

# Compute:
   # history exists?
   # days since last training
   # days since first training
   # fill missing values (farmers with no prior history)

def finalize_history_features(df):

    df = df.copy()

    # if no prior match, these will be NaN
    df["has_prior_history"] = df["prior_n_events_before"].notna().astype(int)

    # replace mssing history with 0 training before
    df["prior_n_events_before"] = df["prior_n_events_before"].fillna(0)

    # days since first/last
    df["prior_days_since_first"] = (df["training_day"] - df["prior_first_day"]).dt.days
    df["prior_days_since_last"] = (df["training_day"] - df["prior_prev_day"]).dt.days

    # fill missing with -1 meaning "no history"
    df["prior_days_since_first"] = df["prior_days_since_first"].fillna(-1)
    df["prior_days_since_last"] = df["prior_days_since_last"].fillna(-1)

    # fill missing adoption rates with 0 (no evidence of adoption)
    rate_cols = [c for c in df.columns if c.startswith("prior_rate_")]
    for c in rate_cols:
        df[c] = df[c].fillna(0)

    return df

train_merged = finalize_history_features(train_merged)
test_merged = finalize_history_features(test_merged)


train_merged[["farmer_name","training_day","has_prior_history","prior_n_events_before","prior_days_since_last"]].head()

farmer_name training_day  has_prior_history  prior_n_events_before  \
0  FAR_aguegon   2024-01-03                  0                    0.0   
1  FAR_eqbhscj   2024-01-03                  0                    0.0   
2  FAR_ezoielz   2024-01-03                  0                    0.0   
3  FAR_hopvqbq   2024-01-03                  0                    0.0   
4  FAR_inwzbie   2024-01-03                  0                    0.0   

   prior_days_since_last  
0                   -1.0  
1                   -1.0  
2                   -1.0  
3                   -1.0  
4                   -1.0

**Add trainer familiarity features (seen trainer before)**

We want:
- has this farmer seen today’s trainer before?
- how many times?

In [33]:
# Build cumulative counts per (farmer, trainer) in prior

pt = prior[["farmer_name", "training_day", "trainer_primary"]].copy()
pt = pt.sort_values(["farmer_name", "trainer_primary", "training_day"])

pt["trainer_count_cum"] = pt.groupby(["farmer_name", "trainer_primary"]).cumcount() + 1
pt["trainer_count_before"] = pt.groupby(["farmer_name", "trainer_primary"])["trainer_count_cum"].shift(1)

pt_for_merge = pt[["farmer_name", "trainer_primary", "training_day", "trainer_count_before"]].sort_values(
    ["farmer_name", "trainer_primary", "training_day"]
)


In [34]:
# Merge trainer counts into train/test using merge_asof

# Ensure datetimes
train_merged["training_day"] = pd.to_datetime(train_merged["training_day"], errors="coerce")
test_merged["training_day"]  = pd.to_datetime(test_merged["training_day"], errors="coerce")
pt_for_merge["training_day"] = pd.to_datetime(pt_for_merge["training_day"], errors="coerce")

# Sort RIGHT keys correctly (time first)
pt_for_merge_sorted = pt_for_merge.sort_values(
    ["training_day", "farmer_name", "trainer_primary"]
).reset_index(drop=True)

# Sort LEFT keys correctly (time first)
train_sorted = train_merged.sort_values(
    ["training_day", "farmer_name", "trainer_primary"]
).reset_index(drop=True)

test_sorted = test_merged.sort_values(
    ["training_day", "farmer_name", "trainer_primary"]
).reset_index(drop=True)

# Merge (leakage-safe)
train_merged = pd.merge_asof(
    train_sorted,
    pt_for_merge_sorted,
    by=["farmer_name", "trainer_primary"],
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)

test_merged = pd.merge_asof(
    test_sorted,
    pt_for_merge_sorted,
    by=["farmer_name", "trainer_primary"],
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)

# Finalize columns
train_merged["trainer_count_before"] = train_merged["trainer_count_before"].fillna(0)
train_merged["seen_trainer_before"]  = (train_merged["trainer_count_before"] > 0).astype(int)

test_merged["trainer_count_before"] = test_merged["trainer_count_before"].fillna(0)
test_merged["seen_trainer_before"]  = (test_merged["trainer_count_before"] > 0).astype(int)


In [35]:
# Finalize trainer familiarity fields

for df in [train_merged, test_merged]:
    df["trainer_count_before"] = df["trainer_count_before"].fillna(0)
    df["seen_trainer_before"]  = (df["trainer_count_before"] > 0).astype(int)


**Topics familiarity (overlap with past topics) using topic multi-hot columns**

use topic multi-hot columns to compute:

- “Has the farmer seen each topic before?”

- “How many of today’s topics were seen before?”

In [36]:
# Get the topic columns list

topic_cols = [c for c in train.columns if c.startswith("topic_")]
print("Topic names: ", topic_cols)
print("Number of topic columns: ", len(topic_cols))

Topic names:  ['topic_diversity', 'topic__The Benefits Of Ndume App', 'topic__Herd Health Management.', 'topic__Poultry Management Practises.', 'topic__Poultry Health Management.', 'topic__Poultry Management Practices', 'topic__Biosecurity In Poultry Farming', 'topic__Feeding A Lactating Cow', 'topic__Calf Feeding', 'topic__How To Succeed In Breeding', 'topic__Reasons Why Ai Fails And Solutions', 'topic__Poultry Management Practises..', 'topic__Poultry Mngt Practices', 'topic__Poultry Health Management. Practices', 'topic__Aflatoxin In Dairy Farming', 'topic__Dairy Feed Management', 'topic__Antimicrobial Resistance', 'topic__Herd Health', 'topic__Record Keeping In Poultry', 'topic__Herd Management', 'topic__Poultry Health Management..', 'topic__Importance Of Mineral Supplementation', 'topic__Poultry Diseases And Biosecurity', 'topic__Pest And Disease Management In Crops', 'topic__Poultry And Dairy Feeding With Tyari Feeds', 'topic__Milking Hygiene', 'topic__Poultry Health Management', 

In [37]:
# Create cumulative topic counts per farmer in prior
# treat each topic column as 0/1 per prior event, then cum-sum over time.

prior_topics = prior[["farmer_name", "training_day"] + topic_cols].copy()
prior_topics = prior_topics.sort_values(["farmer_name","training_day"])

# cumulative count per topic
for c in topic_cols:
    prior_topics[c] = prior_topics.groupby("farmer_name")[c].cumsum()

# shift so it's "before current event"
prior_topics_shift = prior_topics.copy()
prior_topics_shift[topic_cols] = prior_topics_shift.groupby("farmer_name")[topic_cols].shift(1)

prior_topics_shift = prior_topics_shift.sort_values(["farmer_name","training_day"])



In [38]:
# Build a renamed version of prior_topics_shift to avoid all collisions forever.

# Identify topic columns (same ones created earlier)
topic_cols = [c for c in train.columns if c.startswith("topic__")]

# Create a copy of prior_topics_shift and rename its topic columns
prior_topics_shift_renamed = prior_topics_shift.copy()
prior_topics_shift_renamed = prior_topics_shift_renamed.rename(
    columns={c: f"priorcount__{c}" for c in topic_cols}
)

# Sort properly for merge_asof
prior_topics_shift_renamed = prior_topics_shift_renamed.sort_values(
    ["training_day", "farmer_name"]
).reset_index(drop=True)

In [39]:
# Remove old priorcount columns if they already exist

# Drop any previous priorcount__topic__ columns if they exist in train_merged/test_merged
old_priorcount_cols_train = [c for c in train_merged.columns if c.startswith("priorcount__topic__")]
old_priorcount_cols_test  = [c for c in test_merged.columns if c.startswith("priorcount__topic__")]

train_merged = train_merged.drop(columns=old_priorcount_cols_train, errors="ignore")
test_merged  = test_merged.drop(columns=old_priorcount_cols_test, errors="ignore")


In [40]:
# Merge WITHOUT suffixes (no collisions)

# Sort left
train_merged_sorted = train_merged.sort_values(["training_day", "farmer_name"]).reset_index(drop=True)
test_merged_sorted  = test_merged.sort_values(["training_day", "farmer_name"]).reset_index(drop=True)

# Merge
train_merged = pd.merge_asof(
    train_merged_sorted,
    prior_topics_shift_renamed[["training_day","farmer_name"] + [f"priorcount__{c}" for c in topic_cols]],
    by="farmer_name",
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)

test_merged = pd.merge_asof(
    test_merged_sorted,
    prior_topics_shift_renamed[["training_day","farmer_name"] + [f"priorcount__{c}" for c in topic_cols]],
    by="farmer_name",
    on="training_day",
    direction="backward",
    allow_exact_matches=False
)


In [41]:
# Compute overlap features
# We want overlap only for topics present today

def add_topic_overlap_features(df, topic_cols):
    df = df.copy()

    today = df[topic_cols].fillna(0).astype(int)
    prior_seen = (df[[f"priorcount__{c}" for c in topic_cols]].fillna(0) > 0).astype(int)

    overlap = (today * prior_seen).sum(axis=1)
    n_today = today.sum(axis=1)

    df["prior_topics_seen_count"] = overlap
    df["prior_topics_seen_ratio"] = (overlap / n_today.replace(0, np.nan)).fillna(0.0)

    return df

train_merged = add_topic_overlap_features(train_merged, topic_cols)
test_merged  = add_topic_overlap_features(test_merged, topic_cols)



**Quick sanity checks (make sure no leakage)**

In [42]:
# 1) Farmers with no prior history should have 0 events

train_merged["has_prior_history"].value_counts()
train_merged.loc[train_merged["has_prior_history"]==0, "prior_n_events_before"].unique()[:10]

array([0.])

In [43]:
# 2) Days since last should be -1 if no history

train_merged.loc[train_merged["has_prior_history"]==0, "prior_days_since_last"].value_counts().head()

prior_days_since_last
-1.0    10952
Name: count, dtype: int64

In [44]:
# 3) Inspect a random farmer timeline

f = train_merged["farmer_name"].sample(1, random_state=42).iloc[0]
cols_show = ["farmer_name","training_day","prior_n_events_before","prior_days_since_last","seen_trainer_before","trainer_count_before"]
train_merged.loc[train_merged["farmer_name"]==f, cols_show].sort_values("training_day").head(10)


farmer_name training_day  prior_n_events_before  prior_days_since_last  \
2117  FAR_zayuwqd   2024-03-09                    0.0                   -1.0   

      seen_trainer_before  trainer_count_before  
2117                    0                   0.0

### Train 3 models

- train_merged (train + prior history + trainer familiarity + topic overlap)

- test_merged (test + same engineered features)

**Train 3 separate binary classifiers:**

- 07 days

- 90 days

- 120 days

**Then produce the 6 required prediction columns:**

- TX_07_AUC, TX_07_LogLoss

- TX_90_AUC, TX_90_LogLoss

- TX_120_AUC, TX_120_LogLoss

In [45]:
# Imports

from catboost import CatBoostClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score, log_loss

In [46]:
# Define targets and basic checks

TARGETS = {
    "07": "adopted_within_07_days",
    "90": "adopted_within_90_days",
    "120": "adopted_within_120_days"
}

for k, col in TARGETS.items():
    print(k, col, "exists in train_merged:", col in train_merged.columns)

# Ensures target columns exist (prevents silent errors)

07 adopted_within_07_days exists in train_merged: True
90 adopted_within_90_days exists in train_merged: True
120 adopted_within_120_days exists in train_merged: True


In [47]:
# Choose feature columns 

# Columns to always exclude

exclude_cols = set([
    "ID",
    "farmer_name",     # identifier (leaks/memorizes)
    "trainer",         # raw messy col
    "trainer_list",    # list type
    "topics_list",     # raw string list
    "topics_parsed"    # list type
])

exclude_cols.update(TARGETS.values())  # remove target columns

In [48]:
# Build feature list automatically

feature_cols = [c for c in train_merged.columns if c not in exclude_cols]

print("Total features:", len(feature_cols))
print("Feature names:", feature_cols)

Total features: 128
Feature names: ['training_day', 'gender', 'registration', 'age', 'group_name', 'belong_to_cooperative', 'county', 'subcounty', 'ward', 'has_topic_trained_on', 'trainer_primary', 'n_trainers', 'n_topics', 'n_unique_topics', 'topic_diversity', 'topic__The Benefits Of Ndume App', 'topic__Herd Health Management.', 'topic__Poultry Management Practises.', 'topic__Poultry Health Management.', 'topic__Poultry Management Practices', 'topic__Biosecurity In Poultry Farming', 'topic__Feeding A Lactating Cow', 'topic__Calf Feeding', 'topic__How To Succeed In Breeding', 'topic__Reasons Why Ai Fails And Solutions', 'topic__Poultry Management Practises..', 'topic__Poultry Mngt Practices', 'topic__Poultry Health Management. Practices', 'topic__Aflatoxin In Dairy Farming', 'topic__Dairy Feed Management', 'topic__Antimicrobial Resistance', 'topic__Herd Health', 'topic__Record Keeping In Poultry', 'topic__Herd Management', 'topic__Poultry Health Management..', 'topic__Importance Of Min

In [49]:
X = train_merged[feature_cols].copy()
X_test = test_merged[feature_cols].copy()

cat_cols = X.select_dtypes(include=["object", "string", "category"]).columns.tolist()

cat_features = [feature_cols.index(c) for c in cat_cols]

print("Number of categorical features:", len(cat_features))
print("Categorical feature names:", cat_cols)


Number of categorical features: 8
Categorical feature names: ['gender', 'registration', 'age', 'group_name', 'county', 'subcounty', 'ward', 'trainer_primary']


In [50]:
# Remove datetime columns (CatBoost can't use raw datetime)
datetime_cols = X.select_dtypes(include=["datetime"]).columns.tolist()
feature_cols = [c for c in feature_cols if c not in datetime_cols]

X = train_merged[feature_cols].copy()
X_test = test_merged[feature_cols].copy()

# Detect categorical columns 
cat_cols = X.select_dtypes(include=["object", "string", "category"]).columns.tolist()
cat_features = [X.columns.get_loc(c) for c in cat_cols]

print("Number of categorical features:", len(cat_features))
print("Categorical feature names:", cat_cols)


Number of categorical features: 8
Categorical feature names: ['gender', 'registration', 'age', 'group_name', 'county', 'subcounty', 'ward', 'trainer_primary']


In [51]:
# Build a reusable CV training function

# This function will:
# train using 5-fold stratified CV
# output out-of-fold predictions (OOF)
# report AUC and LogLoss
# train a final model on full data for test predictions

def train_catboost_cv(X, y, X_test, cat_features, seed=42, n_splits=5):
    skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=seed)

    oof = np.zeros(len(X))
    test_pred = np.zeros(len(X_test))

    for fold, (tr_idx, val_idx) in enumerate(skf.split(X, y), 1):
        X_tr, X_val = X.iloc[tr_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[tr_idx], y.iloc[val_idx]

        model = CatBoostClassifier(
            iterations=4000,
            learning_rate=0.02,
            depth=6,
            loss_function="Logloss",
            eval_metric="Logloss",
            random_seed=seed,
            verbose=200,
            early_stopping_rounds=300,
            l2_leaf_reg=5
        )

        model.fit(
            X_tr, y_tr,
            cat_features=cat_features,
            eval_set=(X_val, y_val),
            use_best_model=True
        )

        # Predict probabilities
        oof[val_idx] = model.predict_proba(X_val)[:, 1]
        test_pred += model.predict_proba(X_test)[:, 1] / n_splits

        print(f"Fold {fold} done.")

    # Metrics
    auc = roc_auc_score(y, oof)
    ll = log_loss(y, np.clip(oof, 1e-6, 1-1e-6))

    return oof, test_pred, auc, ll


# Why these settings?
# low learning_rate + more iterations helps rare-event learning
# early stopping prevents overfitting
# l2_leaf_reg adds regularization (important with few positives)

In [52]:
# Train 3 models and store results

results = {}
test_preds = {}

for horizon, target_col in TARGETS.items():
    print("\n==============================")
    print("Training target:", horizon, target_col)
    y = train_merged[target_col].astype(int)

    oof, pred_test, auc, ll = train_catboost_cv(X, y, X_test, cat_features)

    results[horizon] = {"auc": auc, "logloss": ll}
    test_preds[horizon] = pred_test

    print(f"Target {horizon} -> AUC={auc:.5f}, LogLoss={ll:.5f}")



Training target: 07 adopted_within_07_days
0:	learn: 0.6491574	test: 0.6492048	best: 0.6492048 (0)	total: 184ms	remaining: 12m 13s
200:	learn: 0.0301811	test: 0.0276119	best: 0.0276119 (200)	total: 18.5s	remaining: 5m 50s
400:	learn: 0.0241456	test: 0.0236841	best: 0.0236841 (400)	total: 32.6s	remaining: 4m 52s
600:	learn: 0.0214428	test: 0.0222486	best: 0.0222486 (600)	total: 43.9s	remaining: 4m 8s
800:	learn: 0.0200018	test: 0.0213450	best: 0.0213439 (798)	total: 54.7s	remaining: 3m 38s
1000:	learn: 0.0181458	test: 0.0201381	best: 0.0201381 (1000)	total: 1m 6s	remaining: 3m 20s
1200:	learn: 0.0169305	test: 0.0194983	best: 0.0194983 (1200)	total: 1m 19s	remaining: 3m 4s
1400:	learn: 0.0157761	test: 0.0187883	best: 0.0187883 (1400)	total: 1m 31s	remaining: 2m 49s
1600:	learn: 0.0144575	test: 0.0185852	best: 0.0185645 (1499)	total: 1m 43s	remaining: 2m 34s
1800:	learn: 0.0134941	test: 0.0184461	best: 0.0184380 (1788)	total: 1m 55s	remaining: 2m 21s
2000:	learn: 0.0128337	test: 0.018109

In [53]:
# Output two probability versions per target
# We will use:
    # AUC column: raw model probabilities (good ranking)
    # LogLoss column: slightly “smoothed” probabilities (reduces extreme confidence)


# define function for a simple and safe smoothing:
def smooth_probs(p, alpha=0.05):
    """
    Shrinks probabilities slightly toward the overall mean.
    This improves logloss stability for very imbalanced targets.
    """
    p = np.asarray(p)
    p_mean = p.mean()
    return (1 - alpha) * p + alpha * p_mean


In [54]:
# Now build submission predictions:

preds_auc = {}
preds_ll = {}

for horizon in TARGETS.keys():
    p = test_preds[horizon]
    preds_auc[horizon] = np.clip(p, 1e-6, 1-1e-6)
    preds_ll[horizon]  = np.clip(smooth_probs(p, alpha=0.05), 1e-6, 1-1e-6)


In [60]:
# Build the submission file

sub = pd.DataFrame({
    "ID": test_merged["ID"],  # ensure this exists
    "Target_07_AUC": preds_auc["07"],
    "Target_90_AUC": preds_auc["90"],
    "Target_120_AUC": preds_auc["120"],
    "Target_07_LogLoss": preds_ll["07"],
    "Target_90_LogLoss": preds_ll["90"],
    "Target_120_LogLoss": preds_ll["120"],
})

sub.head()


ID  Target_07_AUC  Target_90_AUC  Target_120_AUC  Target_07_LogLoss  \
0  ID_PZPP2T       0.002115       0.002045        0.001757           0.002774   
1  ID_F6RQR5       0.001887       0.002461        0.001955           0.002558   
2  ID_E04V5D       0.001887       0.002461        0.001955           0.002558   
3  ID_HXXXEO       0.002187       0.002418        0.001923           0.002843   
4  ID_U9Z9PF       0.004621       0.007554        0.006517           0.005155   

   Target_90_LogLoss  Target_120_LogLoss  
0           0.002871            0.002772  
1           0.003266            0.002960  
2           0.003266            0.002960  
3           0.003225            0.002929  
4           0.008104            0.007294

In [61]:
# Save submission

sub.to_csv("submission_2.csv", index=False)
print("Saved submission.csv")


Saved submission.csv


**Sanity checks for submission**

In [57]:
# 1) No missing values

print(sub.isna().sum())


ID                0
TX_07_AUC         0
TX_07_LogLoss     0
TX_90_AUC         0
TX_90_LogLoss     0
TX_120_AUC        0
TX_120_LogLoss    0
dtype: int64


In [58]:
# 2) All probabilities between 0 and 1

for c in sub.columns:
    if c != "ID":
        print(c, sub[c].min(), sub[c].max())


TX_07_AUC 0.00025302519947779294 0.8929547739425743
TX_07_LogLoss 0.0010051874895307554 0.8490718487954724
TX_90_AUC 0.00010186993887972967 0.9409065069202955
TX_90_LogLoss 0.0010247106092915591 0.8947891157416366
TX_120_AUC 6.478581312905791e-05 0.9557779423197945
TX_120_LogLoss 0.0011644411923978023 0.9090919398737299


In [59]:
# 3) Basic distribution check

sub.describe()

TX_07_AUC  TX_07_LogLoss    TX_90_AUC  TX_90_LogLoss   TX_120_AUC  \
count  5621.000000    5621.000000  5621.000000    5621.000000  5621.000000   
mean      0.015296       0.015296     0.018559       0.018559     0.022058   
std       0.071899       0.068304     0.080605       0.076575     0.097156   
min       0.000253       0.001005     0.000102       0.001025     0.000065   
25%       0.001696       0.002376     0.001880       0.002714     0.001442   
50%       0.002875       0.003496     0.003695       0.004438     0.003581   
75%       0.005291       0.005791     0.007092       0.007666     0.007866   
max       0.892955       0.849072     0.940907       0.894789     0.955778   

       TX_120_LogLoss  
count     5621.000000  
mean         0.022058  
std          0.092299  
min          0.001164  
25%          0.002473  
50%          0.004505  
75%          0.008575  
max          0.909092

## Calibration + dual prediction strategy

In [64]:
# Confirm we are training and predicting on the right columns

TARGETS = {
    "07": "adopted_within_07_days",
    "90": "adopted_within_90_days",
    "120": "adopted_within_120_days",
}

# 1) Ensure targets are not in features
bad = [c for c in feature_cols if c in TARGETS.values()]
print("Targets inside feature_cols:", bad)

# 2) Ensure IDs are not in features
print("ID in features?", "ID" in feature_cols)
print("farmer_name in features?", "farmer_name" in feature_cols)

# 3) Ensure no raw list columns are in features
bad_like = [c for c in feature_cols if "topics" in c.lower() and "topic__" not in c]
print("Suspicious topic cols:", bad_like)

# 4) Ensure no datetime columns in features
dt_cols = train_merged[feature_cols].select_dtypes(include=["datetime"]).columns.tolist()
print("Datetime in features:", dt_cols)

Targets inside feature_cols: []
ID in features? False
farmer_name in features? False
Suspicious topic cols: ['n_topics', 'n_unique_topics', 'prior_topics_seen_count', 'prior_topics_seen_ratio']
Datetime in features: []


In [ ]:
# Check repetition

rows = len(train_merged)
unique_farmers = train_merged["farmer_name"].nunique()
print("Rows:", rows)
print("Unique farmers:", unique_farmers)
print("Avg rows per farmer:", rows / unique_farmers)


Rows: 13536
Unique farmers: 13536
Avg rows per farmer: 1.0


In [66]:
# Use GroupKFold by farmer_name

from sklearn.model_selection import GroupKFold

gkf = GroupKFold(n_splits=5)
groups = train_merged["farmer_name"]


### V1 function & Calibration

In [67]:
# Train CatBoost with GroupKFold + save OOF predictions per target

from catboost import CatBoostClassifier
from sklearn.metrics import roc_auc_score, log_loss
import numpy as np

def train_catboost_groupcv(X, y, X_test, cat_features, groups, n_splits=5, seed=42):
    gkf = GroupKFold(n_splits=n_splits)
    oof = np.zeros(len(X))
    test_pred = np.zeros(len(X_test))

    for fold, (tr_idx, val_idx) in enumerate(gkf.split(X, y, groups=groups), 1):
        X_tr, X_val = X.iloc[tr_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[tr_idx], y.iloc[val_idx]

        model = CatBoostClassifier(
            iterations=3000,
            learning_rate=0.03,
            depth=6,
            loss_function="Logloss",
            eval_metric="Logloss",
            random_seed=seed,
            verbose=200,
            early_stopping_rounds=200
        )

        model.fit(
            X_tr, y_tr,
            cat_features=cat_features,
            eval_set=(X_val, y_val),
            use_best_model=True
        )

        oof[val_idx] = model.predict_proba(X_val)[:, 1]
        test_pred += model.predict_proba(X_test)[:, 1] / n_splits

        print(f"Fold {fold} done.")

    auc = roc_auc_score(y, oof)
    ll  = log_loss(y, np.clip(oof, 1e-6, 1-1e-6))
    return oof, test_pred, auc, ll


In [68]:
# Run it for each target

results = {}
test_preds = {}
oof_preds = {}

for horizon, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)

    oof, pred_test, auc, ll = train_catboost_groupcv(
        X, y, X_test, cat_features,
        groups=train_merged["farmer_name"],
        n_splits=5
    )

    results[horizon] = {"auc": auc, "logloss": ll}
    test_preds[horizon] = pred_test
    oof_preds[horizon] = oof

    print(horizon, results[horizon])


0:	learn: 0.6278880	test: 0.6278904	best: 0.6278904 (0)	total: 35ms	remaining: 1m 45s
200:	learn: 0.0244266	test: 0.0220715	best: 0.0220612 (199)	total: 9.81s	remaining: 2m 16s
400:	learn: 0.0191377	test: 0.0198872	best: 0.0198872 (400)	total: 22.8s	remaining: 2m 27s
600:	learn: 0.0151431	test: 0.0184191	best: 0.0184166 (595)	total: 35.6s	remaining: 2m 22s
800:	learn: 0.0129664	test: 0.0178702	best: 0.0178612 (797)	total: 48.1s	remaining: 2m 12s
1000:	learn: 0.0116817	test: 0.0176467	best: 0.0176431 (996)	total: 1m	remaining: 2m 1s
1200:	learn: 0.0102999	test: 0.0176118	best: 0.0175903 (1083)	total: 1m 13s	remaining: 1m 50s
Stopped by overfitting detector  (200 iterations wait)

bestTest = 0.01759026219
bestIteration = 1083

Shrink model to first 1084 iterations.
Fold 1 done.
0:	learn: 0.6287136	test: 0.6286409	best: 0.6286409 (0)	total: 61.5ms	remaining: 3m 4s
200:	learn: 0.0198059	test: 0.0414641	best: 0.0414641 (200)	total: 13.4s	remaining: 3m 6s
400:	learn: 0.0150872	test: 0.041616

In [69]:
# Calibrate for LogLoss


# Use Platt scaling (sigmoid calibration) using OOF predictions

from sklearn.linear_model import LogisticRegression

def fit_platt(oof, y):
    oof = np.clip(oof, 1e-6, 1-1e-6).reshape(-1, 1)
    clf = LogisticRegression(solver="lbfgs", max_iter=2000)
    clf.fit(oof, y)
    return clf

def apply_platt(clf, p):
    p = np.clip(p, 1e-6, 1-1e-6).reshape(-1, 1)
    return clf.predict_proba(p)[:, 1]


# Apply per target:

test_preds_cal = {}

for horizon, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)
    calib = fit_platt(oof_preds[horizon], y)
    test_preds_cal[horizon] = apply_platt(calib, test_preds[horizon])


In [70]:
# Build the submission

sub = pd.DataFrame({"ID": test_merged["ID"]})

sub["Target_07_AUC"]     = np.clip(test_preds["07"], 1e-6, 1-1e-6)
sub["Target_90_AUC"]     = np.clip(test_preds["90"], 1e-6, 1-1e-6)
sub["Target_120_AUC"]     = np.clip(test_preds["120"], 1e-6, 1-1e-6)

sub["TX_07_LogLoss"] = np.clip(test_preds_cal["07"], 1e-6, 1-1e-6)
sub["Target_90_LogLoss"] = np.clip(test_preds_cal["90"], 1e-6, 1-1e-6)
sub["Target_120_LogLoss"] = np.clip(test_preds_cal["120"], 1e-6, 1-1e-6)

sub.to_csv("submission_2_groupcv_calibrated.csv", index=False)
sub.head()


ID  Target_07_AUC  Target_90_AUC  Target_120_AUC  TX_07_LogLoss  \
0  ID_PZPP2T       0.001595       0.001592        0.001824       0.005835   
1  ID_F6RQR5       0.001240       0.001823        0.001905       0.005817   
2  ID_E04V5D       0.001240       0.001823        0.001905       0.005817   
3  ID_HXXXEO       0.001518       0.001761        0.001876       0.005831   
4  ID_U9Z9PF       0.003934       0.006433        0.008050       0.005957   

   Target_90_LogLoss  Target_120_LogLoss  
0           0.006949            0.007677  
1           0.006962            0.007682  
2           0.006962            0.007682  
3           0.006959            0.007680  
4           0.007234            0.008098

### V2 Function & Calibration

In [71]:
from sklearn.model_selection import GroupKFold
from sklearn.metrics import roc_auc_score, log_loss
from catboost import CatBoostClassifier
import numpy as np

def train_catboost_groupcv(X, y, X_test, cat_features, groups, n_splits=5, seed=42):
    gkf = GroupKFold(n_splits=n_splits)
    oof = np.zeros(len(X))
    test_pred = np.zeros(len(X_test))

    for fold, (tr_idx, val_idx) in enumerate(gkf.split(X, y, groups=groups), 1):
        X_tr, X_val = X.iloc[tr_idx], X.iloc[val_idx]
        y_tr, y_val = y.iloc[tr_idx], y.iloc[val_idx]

        model = CatBoostClassifier(
            iterations=2500,
            learning_rate=0.03,
            depth=6,
            loss_function="Logloss",
            eval_metric="Logloss",
            random_seed=seed,
            verbose=200,
            early_stopping_rounds=200,
            l2_leaf_reg=5
        )

        model.fit(
            X_tr, y_tr,
            cat_features=cat_features,
            eval_set=(X_val, y_val),
            use_best_model=True
        )

        oof[val_idx] = model.predict_proba(X_val)[:, 1]
        test_pred += model.predict_proba(X_test)[:, 1] / n_splits

        print(f"Fold {fold} complete.")

    auc = roc_auc_score(y, oof)
    ll  = log_loss(y, np.clip(oof, 1e-6, 1-1e-6))
    return oof, test_pred, auc, ll


In [72]:
# Run it per target

results = {}
test_preds = {}
oof_preds = {}

groups = train_merged["farmer_name"]

for horizon, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)

    oof, pred_test, auc, ll = train_catboost_groupcv(
        X, y, X_test, cat_features,
        groups=groups, n_splits=5
    )

    results[horizon] = {"auc": auc, "logloss": ll}
    test_preds[horizon] = pred_test
    oof_preds[horizon] = oof

    print(horizon, results[horizon])


0:	learn: 0.6283373	test: 0.6283397	best: 0.6283397 (0)	total: 40.6ms	remaining: 1m 41s
200:	learn: 0.0256400	test: 0.0223944	best: 0.0223944 (200)	total: 11.3s	remaining: 2m 8s
400:	learn: 0.0207430	test: 0.0200200	best: 0.0200148 (399)	total: 25.2s	remaining: 2m 11s
600:	learn: 0.0173000	test: 0.0182290	best: 0.0182200 (597)	total: 37.6s	remaining: 1m 58s
800:	learn: 0.0149982	test: 0.0172439	best: 0.0172439 (800)	total: 49.5s	remaining: 1m 44s
1000:	learn: 0.0138351	test: 0.0172280	best: 0.0171277 (912)	total: 1m 2s	remaining: 1m 33s
1200:	learn: 0.0124266	test: 0.0170627	best: 0.0170525 (1198)	total: 1m 16s	remaining: 1m 22s
1400:	learn: 0.0115264	test: 0.0170223	best: 0.0169524 (1331)	total: 1m 28s	remaining: 1m 9s
Stopped by overfitting detector  (200 iterations wait)

bestTest = 0.01695237307
bestIteration = 1331

Shrink model to first 1332 iterations.
Fold 1 complete.
0:	learn: 0.6287579	test: 0.6286546	best: 0.6286546 (0)	total: 51.7ms	remaining: 2m 9s
200:	learn: 0.0213991	te

In [73]:
# Calibrate for LogLoss columns (high impact)
# Platt calibration using OOF predictions:


from sklearn.linear_model import LogisticRegression

def fit_platt(oof, y):
    x = np.clip(oof, 1e-6, 1-1e-6).reshape(-1, 1)
    clf = LogisticRegression(solver="lbfgs", max_iter=2000)
    clf.fit(x, y)
    return clf

def apply_platt(clf, p):
    x = np.clip(p, 1e-6, 1-1e-6).reshape(-1, 1)
    return clf.predict_proba(x)[:, 1]

test_preds_cal = {}
for horizon, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)
    calib = fit_platt(oof_preds[horizon], y)
    test_preds_cal[horizon] = apply_platt(calib, test_preds[horizon])


In [74]:
# Submission (raw for AUC, calibrated for LogLoss)

sub = pd.DataFrame({"ID": test_merged["ID"]})

sub["Target_07_AUC"]     = np.clip(test_preds["07"], 1e-6, 1-1e-6)
sub["Target_90_AUC"]     = np.clip(test_preds["90"], 1e-6, 1-1e-6)
sub["Target_120_AUC"]     = np.clip(test_preds["120"], 1e-6, 1-1e-6)

sub["Target_07_LogLoss"] = np.clip(test_preds_cal["07"], 1e-6, 1-1e-6)
sub["Target_90_LogLoss"] = np.clip(test_preds_cal["90"], 1e-6, 1-1e-6)
sub["Target_120_LogLoss"] = np.clip(test_preds_cal["120"], 1e-6, 1-1e-6)

sub.to_csv("submission_v3_groupcv_platt.csv", index=False)
sub.head()


ID  Target_07_AUC  Target_90_AUC  Target_120_AUC  Target_07_LogLoss  \
0  ID_PZPP2T       0.001418       0.002523        0.001320           0.005814   
1  ID_F6RQR5       0.001295       0.002623        0.001473           0.005808   
2  ID_E04V5D       0.001295       0.002623        0.001473           0.005808   
3  ID_HXXXEO       0.001493       0.002654        0.001494           0.005818   
4  ID_U9Z9PF       0.004300       0.007765        0.008586           0.005961   

   Target_90_LogLoss  Target_120_LogLoss  
0           0.006981            0.007576  
1           0.006987            0.007586  
2           0.006987            0.007586  
3           0.006988            0.007587  
4           0.007301            0.008060

In [75]:
print(len(train_merged)/train_merged["farmer_name"].nunique())

1.0


In [78]:
def smooth_probs(p, alpha=0.05):
    p = np.asarray(p)
    mu = p.mean()
    return (1 - alpha) * p + alpha * mu


In [79]:
sub = pd.DataFrame({"ID": test_merged["ID"]})

for h in ["07","90","120"]:
    sub[f"TX_{h}_AUC"] = np.clip(test_preds[h], 1e-6, 1-1e-6)
    sub[f"TX_{h}_LogLoss"] = np.clip(smooth_probs(test_preds[h], alpha=0.05), 1e-6, 1-1e-6)

sub.to_csv("submission_smooth_005.csv", index=False)


In [81]:
for a in [0.02, 0.05, 0.10, 0.15]:
    sub = pd.DataFrame({"ID": test_merged["ID"]})
    for h in ["07","90","120"]:
        sub[f"Target_{h}_AUC"] = np.clip(test_preds[h], 1e-6, 1-1e-6)
        sub[f"Target_{h}_LogLoss"] = np.clip(smooth_probs(test_preds[h], alpha=a), 1e-6, 1-1e-6)
    sub.to_csv(f"submission_smooth_{a:.2f}.csv", index=False)


## Ensembling

In [84]:
# Create a “clean modeling matrix” for both models
# Why? CatBoost can handle categoricals directly, but LightGBM usually needs encoding

# Build x and x_test from our merged data

X_train = train_merged[feature_cols].copy()
X_test = test_merged[feature_cols].copy()

# Identify categorical columns by dtype

cats_cols = X_train.select_dtypes(include = ["object", "string", "category"]).columns.tolist()
num_cols = [c for c in X_train.columns if c not in cat_cols]

print("Num features:", X_train.shape[1])
print("categorical:", len(cat_cols))
print("Numeric:", len(num_cols))
print("Example cat cols:", cat_cols[:10])
print("Example num cols:", num_cols[:10])

Num features: 125
categorical: 8
Numeric: 117
Example cat cols: ['gender', 'registration', 'age', 'group_name', 'county', 'subcounty', 'ward', 'trainer_primary']
Example num cols: ['belong_to_cooperative', 'has_topic_trained_on', 'n_trainers', 'n_topics', 'n_unique_topics', 'topic_diversity', 'topic__The Benefits Of Ndume App', 'topic__Herd Health Management.', 'topic__Poultry Management Practises.', 'topic__Poultry Health Management.']


In [85]:
# Train a second CatBoost (different config)
    # different depth
    # more regularization
    # different sampling strength

from catboost import CatBoostClassifier
from sklearn.model_selection import StratifiedKFold
from sklearn.metrics import roc_auc_score, log_loss

def catboost_cv_predict(X, y, X_test, cat_cols, params, n_splits=5, seed=42):
    skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=seed)
    oof = np.zeros(len(X))
    test_pred = np.zeros(len(X_test))

    # CatBoost expects categorical feature indices
    cat_features = [X.columns.get_loc(c) for c in cat_cols]

    for fold, (tr_idx, va_idx) in enumerate(skf.split(X, y), 1):
        X_tr, X_va = X.iloc[tr_idx], X.iloc[va_idx]
        y_tr, y_va = y.iloc[tr_idx], y.iloc[va_idx]

        model = CatBoostClassifier(**params)
        model.fit(
            X_tr, y_tr,
            cat_features=cat_features,
            eval_set=(X_va, y_va),
            use_best_model=True,
            verbose=200
        )

        oof[va_idx] = model.predict_proba(X_va)[:, 1]
        test_pred += model.predict_proba(X_test)[:, 1] / n_splits
        print(f"Fold {fold} done.")

    auc = roc_auc_score(y, oof)
    ll = log_loss(y, np.clip(oof, 1e-6, 1-1e-6))
    return oof, test_pred, auc, ll


In [86]:
# A good “complement” CatBoost params

cb_params_2 = dict(
    iterations=4000,
    learning_rate=0.03,
    depth=5,                # different from previous depth=6
    l2_leaf_reg=15,         # more regularization
    random_strength=2.0,
    bagging_temperature=1.0,
    loss_function="Logloss",
    eval_metric="Logloss",
    early_stopping_rounds=250,
    random_seed=42
)

In [ ]:
# Train for each target

cb2_oof = {}
cb2_test = {}
cb2_results = {}

for h, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)
    oof, pred_test, auc, ll = catboost_cv_predict(
        X_train, y, X_test, cat_cols, cb_params_2, n_splits=5, seed=42
    )
    cb2_oof[h] = oof
    cb2_test[h] = pred_test
    cb2_results[h] = {"auc": auc, "logloss": ll}
    print("Target", h, cb2_results[h])


0:	learn: 0.6290280	test: 0.6291200	best: 0.6291200 (0)	total: 192ms	remaining: 12m 46s
200:	learn: 0.0340123	test: 0.0308248	best: 0.0308233 (199)	total: 12.9s	remaining: 4m 3s
400:	learn: 0.0295539	test: 0.0269805	best: 0.0269795 (399)	total: 25.5s	remaining: 3m 49s
600:	learn: 0.0260813	test: 0.0242256	best: 0.0242256 (600)	total: 38.4s	remaining: 3m 37s
800:	learn: 0.0237179	test: 0.0228263	best: 0.0228263 (800)	total: 51.7s	remaining: 3m 26s
1000:	learn: 0.0223630	test: 0.0222070	best: 0.0222070 (1000)	total: 1m 4s	remaining: 3m 13s
1200:	learn: 0.0210983	test: 0.0215855	best: 0.0215796 (1195)	total: 1m 17s	remaining: 2m 59s
1400:	learn: 0.0201912	test: 0.0209692	best: 0.0209672 (1397)	total: 1m 29s	remaining: 2m 46s
1600:	learn: 0.0194095	test: 0.0204917	best: 0.0204914 (1597)	total: 1m 42s	remaining: 2m 33s
1800:	learn: 0.0185214	test: 0.0203173	best: 0.0203173 (1800)	total: 1m 54s	remaining: 2m 20s
2000:	learn: 0.0175112	test: 0.0199559	best: 0.0199493 (1999)	total: 2m 7s	remai

In [88]:
%pip install lightgbm

  Using cached lightgbm-4.6.0-py3-none-win_amd64.whl.metadata (17 kB)
Using cached lightgbm-4.6.0-py3-none-win_amd64.whl (1.5 MB)
Note: you may need to restart the kernel to use updated packages.


In [89]:
# Train LightGBM
# Encode categoricals (LightGBM needs numbers). Use Ordinal encoding (simple + works well with tree models)

from sklearn.preprocessing import OrdinalEncoder

# Fit encoder on combined train+test for consistency
enc = OrdinalEncoder(handle_unknown="use_encoded_value", unknown_value=-1)

X_train_lgb = X_train.copy()
X_test_lgb  = X_test.copy()

X_train_lgb[cat_cols] = enc.fit_transform(X_train_lgb[cat_cols].astype(str))
X_test_lgb[cat_cols]  = enc.transform(X_test_lgb[cat_cols].astype(str))


In [90]:
# Train LightGBM with CV

import lightgbm as lgb

def lgbm_cv_predict(X, y, X_test, params, n_splits=5, seed=42):
    skf = StratifiedKFold(n_splits=n_splits, shuffle=True, random_state=seed)
    oof = np.zeros(len(X))
    test_pred = np.zeros(len(X_test))

    for fold, (tr_idx, va_idx) in enumerate(skf.split(X, y), 1):
        X_tr, X_va = X.iloc[tr_idx], X.iloc[va_idx]
        y_tr, y_va = y.iloc[tr_idx], y.iloc[va_idx]

        model = lgb.LGBMClassifier(**params)
        model.fit(
            X_tr, y_tr,
            eval_set=[(X_va, y_va)],
            eval_metric="binary_logloss",
            callbacks=[lgb.early_stopping(200, verbose=False)]
        )

        oof[va_idx] = model.predict_proba(X_va)[:, 1]
        test_pred += model.predict_proba(X_test)[:, 1] / n_splits
        print(f"Fold {fold} done.")

    auc = roc_auc_score(y, oof)
    ll = log_loss(y, np.clip(oof, 1e-6, 1-1e-6))
    return oof, test_pred, auc, ll


In [94]:
# starting LGBM params for imbalanced data

lgb_params = dict(
    n_estimators=8000,
    learning_rate=0.01,
    num_leaves=128,
    max_depth=-1,
    min_child_samples=10,        # reduce from 50
    subsample=0.8,
    colsample_bytree=0.8,
    reg_alpha=0.0,               # reduce regularization
    reg_lambda=1.0,
    scale_pos_weight= (10706 / 122),  # IMPORTANT
    random_state=42,
    n_jobs=-1
)


In [95]:
# Train per target:

lgb_oof = {}
lgb_test = {}
lgb_results = {}

for h, target_col in TARGETS.items():
    y = train_merged[target_col].astype(int)
    oof, pred_test, auc, ll = lgbm_cv_predict(
        X_train_lgb, y, X_test_lgb, lgb_params, n_splits=5, seed=42
    )
    lgb_oof[h] = oof
    lgb_test[h] = pred_test
    lgb_results[h] = {"auc": auc, "logloss": ll}
    print("Target", h, lgb_results[h])


[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] Number of positive: 122, number of negative: 10706
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.018351 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 1293
[LightGBM] [Info] Number of data points in the train set: 10828, number of used features: 122
[LightGBM] [Warning] Found whitespace in feature_names, replace with underlines
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.011267 -> initscore=-4.474539
[LightGBM] [Info] Start training from score -4.474539
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain,